src/lib/transition_frontier_persistence/transition_frontier_persistence.ml

```diff
@@ -49,57 +66,84 @@ module Make (Inputs : Intf.Main_inputs) = struct
+        | `Continue diff_pairs ->
+            let%bind new_hash =
+              Deferred.List.fold diff_pairs ~init:init_hash
+                ~f:(fun acc_hash
+                   (Transition_frontier.Diff_mutant.E.With_value
+                     (diff, ground_truth_mutant))
+                   ->
+                  write_diff_and_verify ~logger ~acc_hash worker
+                    (diff, ground_truth_mutant) )
+            in
+            process_diff_work new_hash
+      in
+      process_diff_work Transition_frontier.Diff_hash.empty
+    in
+    let t =
+      { worker
+      ; worker_mvar= Mvar.write_only worker_mvar
+      ; flush_capacity
+      ; max_buffer_capacity
+      ; buffer
+      ; stop_signal
+      ; worker_thread }
+    in
+    let stop_thread =
+      let%bind `Stop = Ivar.read stop_signal in
+      let%bind () = flush t in
+      Mvar.put worker_mvar `Finished
+    in
+    don't_wait_for stop_thread ; t
 
-  let flush {writer; buffer; _} =
-    let list = Queue.to_list buffer in
-    Pipe.write_without_pushback writer list ;
-    Queue.clear buffer
+  let close {worker; _} = Worker.close worker
 
-  let close_and_finish_copy ({worker; writer; worker_thread; _} as t) =
-    flush t ;
-    Pipe.close writer ;
+  let close_and_finish_copy {worker; stop_signal; worker_thread; _} =
+    Ivar.fill stop_signal `Stop ;
     let%map () = worker_thread in
     Worker.close worker
 
   let listen_to_frontier_broadcast_pipe
       (frontier_broadcast_pipe :
         Transition_frontier.t option Broadcast_pipe.Reader.t)
-      ({max_buffer_capacity; buffer; _} as t) =
+      ({max_buffer_capacity; flush_capacity; buffer; stop_signal; _} as t) =
     Broadcast_pipe.Reader.iter frontier_broadcast_pipe
       ~f:
         (Option.value_map ~default:Deferred.unit ~f:(fun frontier ->
              Broadcast_pipe.Reader.iter
                (Transition_frontier.persistence_diff_pipe frontier)
                ~f:(fun new_diffs ->
-                 Queue.enqueue_all buffer new_diffs ;
-                 if Queue.length buffer >= max_buffer_capacity then (
-                   flush t ;
-                   assert (Queue.is_empty buffer) ) ;
-                 Deferred.unit ) ))
+                 let enqueue_and_flush_work () =
+                   Queue.enqueue_all buffer new_diffs ;
+                   if Queue.length buffer >= max_buffer_capacity then
+                     (* We would only have failure when we are debugging code  *)
+                     Debug_assert.debug_assert_deferred
+                     @@ fun () ->
+                     failwithf
+                       !"There is too many work that a Transition Frontier \
+                         Persistence worker is waiting for. Retune buffer \
+                         parameters: {flush_capacity: %i, buffer_capacity: %i}"
+                       flush_capacity max_buffer_capacity ()
+                   else if Queue.length buffer >= flush_capacity then flush t
```

Comment:
You should only flush if the current job has finished. With the way this is written right now, you have no way to wait for this even though. The logic for pulling off of the queue should happen after the thread at the other end of the flush has completed its job. With this setup, you are always flushing eagerly, which has the effect of treating the Mvar like a queue, which is not what we want (and also Mvar does not guarantee order or delivery of multiple writes as it is not a queue).